pkg/cloudprovider/aws/securitygroups.go

```diff
@@ -36,26 +39,47 @@ func NewSecurityGroupProvider(ec2api ec2iface.EC2API) *SecurityGroupProvider {
 	}
 }
 
-func (s *SecurityGroupProvider) Get(ctx context.Context, clusterName string) ([]*ec2.SecurityGroup, error) {
-	if securityGroups, ok := s.cache.Get(clusterName); ok {
-		return securityGroups.([]*ec2.SecurityGroup), nil
+func (s *SecurityGroupProvider) Get(ctx context.Context, provisioner *v1alpha1.Provisioner, constraints *Constraints) ([]*ec2.SecurityGroup, error) {
+	// 1. Get Security Groups
+	securityGroups, err := s.getSecurityGroups(ctx, provisioner.Spec.Cluster.Name)
+	if err != nil {
+		return nil, err
 	}
-	return s.getSecurityGroups(ctx, clusterName)
+	// 2. Filter by subnet name if constrained
+	if name := constraints.GetSecurityGroupName(); name != nil {
+		securityGroups = filterSecurityGroups(utils.ByName(aws.StringValue(name)), securityGroups)
+	}
+	// 3. Filter by security group tag key if constrained
+	if name := constraints.GetSecurityGroupTagKey(); name != nil {
+		securityGroups = filterSecurityGroups(utils.ByTagKey(aws.StringValue(name)), securityGroups)
+	}
+	return securityGroups, nil
 }
 
 func (s *SecurityGroupProvider) getSecurityGroups(ctx context.Context, clusterName string) ([]*ec2.SecurityGroup, error) {
-	describeSecurityGroupOutput, err := s.ec2api.DescribeSecurityGroupsWithContext(ctx, &ec2.DescribeSecurityGroupsInput{
+	if securityGroups, ok := s.cache.Get(clusterName); ok {
+		return securityGroups.([]*ec2.SecurityGroup), nil
+	}
+	output, err := s.ec2api.DescribeSecurityGroupsWithContext(ctx, &ec2.DescribeSecurityGroupsInput{
 		Filters: []*ec2.Filter{{
-			Name:   aws.String("tag-key"),
+			Name:   aws.String("tag-key"), // Security Groups must be tagged for the cluster
 			Values: []*string{aws.String(fmt.Sprintf(ClusterTagKeyFormat, clusterName))},
 		}},
 	})
 	if err != nil {
 		return nil, fmt.Errorf("describing security groups with tag key %s, %w", fmt.Sprintf(ClusterTagKeyFormat, clusterName), err)
 	}
+	s.cache.Set(clusterName, output.SecurityGroups, CacheTTL)
```

Comment:
Here is a scenario for a customer use case-

- Installed Karpenter, created a pending pods, Karpenter brings up an instance
- User checks the security group doesn't like it, reads the documentation, can create their own security group
- User creates a new security group, adds the SG name to pod and re-creates the pod

Will we not see the new security group until cacheTTL expires?

Reply:
Correct, this will fail (and loop) until the security group becomes visible and it eventually heals.

Reply:
We still need to filter by subnet name and group tag key after we call the AWS API though, right?